crates/hir-def/src/nameres/path_resolution.rs

```diff
@@ -93,11 +93,15 @@ impl DefMap {
                 if remaining.is_some() {
                     return None;
                 }
-                let types = result.take_types()?;
-                match types {
-                    ModuleDefId::ModuleId(m) => Visibility::Module(m),
-                    _ => {
-                        // error: visibility needs to refer to module
+                let types = result.take_types();
+
+                match (types, path.kind) {
+                    (Some(ModuleDefId::ModuleId(m)), _) => Visibility::Module(m),
+                    // resolve_path doesn't find any values for a plan pathkind of a private function
+                    (None, PathKind::Plain | PathKind::Crate) => {
+                        Visibility::Module(self.module_id(original_module))
+                    }
```

Comment:
This is wrong, the issue seems to be that for some reason we construct a plain path with 0 segments for `pub(self)`, while we should be creating a `Super(0)` path with 0 segments

Reply:
Thanks for confirming my suspicion that the root cause was in nameres

> Please let me know what's the best way to add a unit tests to nameres

for the self case as well as the `pub(in crate::krate)` case

Reply:
is there a difference between `pub(self)` and `pub(self::)`, because they end up with different visibility values

Reply:
I think the problem happens in item tree construction, so a place to add a test case would be in `hir-def/src/item_tree/tests.rs` (ideally one that tests all kinds of visibility paths). `pub(self::)` is different in that this contains an incomplete path, I think they ideally should have the same visibility. Might be good to check what happens here in case of other paths (that is `super` vs `super::`, `crate` vs `crate::` etc_.